anndata/_core/merge.py

```diff
@@ -141,7 +141,21 @@ def equal_dask_array(a, b) -> bool:
 
 @equal.register(np.ndarray)
 def equal_array(a, b) -> bool:
-    return equal(pd.DataFrame(a), pd.DataFrame(asarray(b)))
+    def _is_numeric_or_bool(x) -> bool:
+        """array_equal with equal_nan=True only works for these types."""
+        return np.issubdtype(x, np.number) or np.issubdtype(x, np.bool_)
+
+    b = asarray(b)
+
+    if a.shape != b.shape:
+        return False
+
+    if _is_numeric_or_bool(a.dtype) and _is_numeric_or_bool(b.dtype):
+        return np.array_equal(a, b, equal_nan=True)
```

Comment:
Why introduce `array_equal` at all?  I think `reshape` is generally very quick (returns a view of the same underlying in-memory data etc.), so is the array equality much faster here than pandas?

Reply:
So my initial thinking was that sometimes `reshape` and creating the dataframe copies. But I tested a few things at the repl and found that pandas was generally faster here, so I've made the change.